qiskit/providers/aer/openpulse/hamiltonian_model.py

```diff
@@ -0,0 +1,313 @@
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+# pylint: disable=eval-used, exec-used, invalid-name
+
+"PulseModel class for PulseSimulator"
+
+from collections import OrderedDict
+import numpy as np
+import numpy.linalg as la
+from .qobj.opparse import HamiltonianParser
+# pylint: disable=no-name-in-module,import-error
+from .qobj import op_qobj as op
+
+
+class HamiltonianModel():
+    """Hamiltonian model for pulse simulator."""
+    def __init__(self, hamiltonian, qubits=None):
+        """Initialize a Hamiltonian model.
+
+        Args:
+            hamiltonian (dict): Hamiltonian dictionary.
+            qubits (list or None): List of qubits to extract from the hamiltonian.
+
+        Raises:
+            ValueError: if arguments are invalid.
+        """
+        # Initialize internal variables
+        # The system Hamiltonian in numerical format
+        self._system = None
+        # System variables
+        self._vars = None
+        # Channels in the Hamiltonian string
+        # These tell the order in which the channels are evaluated in
+        # the RHS solver.
+        self._channels = None
+        # Diagonal elements of the hamiltonian
+        self._h_diag = None
+        # Eigenvalues of the time-independent hamiltonian
+        self._evals = None
+        # Eigenstates of the time-indepedent hamiltonian
+        self._estates = None
+        # Qubit subspace dimensinos
+        self._dim_qub = {}
+        # Oscillator subspace dimensions
+        self._dim_osc = {}
+
+        # Parse Hamiltonian
+        # TODO: determine n_qubits from hamiltonian if qubits is None
+        n_qubits = len(qubits) if qubits else None
+        if not n_qubits:
+            raise ValueError("TODO: Need to infer n_qubits from "
+                             "Hamiltonian if qubits list is not specified")
+
+        self._vars = OrderedDict(hamiltonian['vars'])
+
+        # Get qubit subspace dimensions
+        if 'qub' in hamiltonian:
+            self._dim_qub = {
+                int(key): val
+                for key, val in hamiltonian['qub'].items()
+            }
+        else:
+            self._dim_qub = {}.fromkeys(range(n_qubits), 2)
+
+        # Get oscillator subspace dimensions
+        if 'osc' in hamiltonian:
+            self._dim_osc = {
+                int(key): val
+                for key, val in hamiltonian['osc'].items()
+            }
+
+        # Step 1: Parse the Hamiltonian
+        system = HamiltonianParser(h_str=hamiltonian['h_str'],
+                                   dim_osc=self._dim_osc,
+                                   dim_qub=self._dim_qub)
+        system.parse(qubits)
+        self._system = system.compiled
+
+        # Step #2: Determine Hamiltonian channels
+        self._get_hamiltonian_channels()
+
+        # Step 3: Calculate diagonal hamiltonian
+        self._get_diag_hamiltonian()
+
+    def initial_state(self):
```

Comment:
Along the lines of what you had suggested, I think a good name/argument for this method would be either `eigenstate(level=0)` or even `drift_eigenstate(level=0)`. It's a bit more verbose but `drift_eigenstate` is very descriptive.